### PR TITLE
Add local cache for Parquet Metadata

### LINKED ISSUE #13921: Provide optional cache for Parquet file metadata
Similar to the ORC changes that introduce metadata caching in https://github.com/prestodb/presto/pull/13501 and https://github.com/prestodb/presto/pull/13575 it would be beneficial to provide this mechanism for Parquet files as well to potentially avoid 2 extra reads per file (tail read for metadata length + actual metadata read) and avoid remote reads completely in subsequent executions where min/max stats can rule out the file immediately.

### PR DESCRIPTION
**Pull request by ClarenceThreepwood**
Test plan - TravisCI and unit tests in AbstractTestParquetReader

Fixes #13921

```
== RELEASE NOTES ==

Hive Changes
* Add caching module for Parquet metadata
```

### CODE REVIEW

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 1]
```diff
@@ -293,4 +299,23 @@ public StripeMetadataSource createStripeMetadataSource(OrcCacheConfig orcCacheCo
         }
         return stripeMetadataSource;
     }
+
+    @Singleton
+    @Provides
+    public ParquetMetadataSource createParquetMetadataSource(ParquetCacheConfig parquetCacheConfig, MBeanExporter exporter)
+    {
+        ParquetMetadataSource parquetMetadataSource = new MetadataReader();
+        if (parquetCacheConfig.ismetadataCacheEnabled()) {
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:308]: nit: `isMetadataCacheEnabled()`

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 2]
```diff
@@ -293,4 +299,23 @@ public StripeMetadataSource createStripeMetadataSource(OrcCacheConfig orcCacheCo
...
+            Cache<ParquetDataSourceId, ParquetFileMetadata> cache = CacheBuilder.newBuilder()
+                    .maximumWeight(parquetCacheConfig.getmetadataCacheSize().toBytes())
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:310]: nit: getMetadataCacheSize()

File: presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java [hunk 3]
```diff
@@ -293,4 +299,23 @@ public StripeMetadataSource createStripeMetadataSource(OrcCacheConfig orcCacheCo
...
+                    .weigher((id, metadata) -> ((ParquetFileMetadata) metadata).getMetadataSize())
+                    .expireAfterAccess(parquetCacheConfig.getmetadataCacheTtlSinceLastAccess().toMillis(), TimeUnit.MILLISECONDS)
```
Comment (shixuan-fan) [inline presto-hive/src/main/java/com/facebook/presto/hive/HiveClientModule.java:312]: nit: `getMetadataCacheTtlSinceLastAccess`

File: presto-parquet/src/main/java/com/facebook/presto/parquet/cache/ParquetCacheConfig.java [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    private Duration metadataCacheTtlSinceLastAccess = new Duration(0, SECONDS);
+
+    public boolean ismetadataCacheEnabled()
+    {
+        return metadataCacheEnabled;
+    }
+
+    @Config("parquet.metadata-cache-enabled")
+    @ConfigDescription("Enable cache for parquet metadata")
+    public ParquetCacheConfig setmetadataCacheEnabled(boolean metadataCacheEnabled)
```
Comment (shixuan-fan) [inline presto-parquet/src/main/java/com/facebook/presto/parquet/cache/ParquetCacheConfig.java:39]: nit: Capitalize the first letter after `get` and `set` in this class

**Review by shixuan-fan — COMMENTED** [review]
Minor nits. LGTM. @zhenxiao could you also take a quick pass since you are more familiar with Parquet? :)

**Review by zhenxiao — APPROVED** [review]
nice work, @ClarenceThreepwood 
LGTM, just one minor thing. please address @shixuan-fan 's comments

File: presto-hive/src/test/java/com/facebook/presto/hive/parquet/AbstractTestParquetReader.java [hunk 5]
```diff
@@ -1517,6 +1532,7 @@ private void setParquetLogging()
     {
         Logger.getLogger(ParquetOutputFormat.class.getName()).setLevel(Level.WARNING);
         Logger.getLogger(CodecConfig.class.getName()).setLevel(Level.WARNING);
+        Logger.getLogger(CodecConfig.class.getName()).setLevel(Level.WARNING);
```
Comment (zhenxiao) [inline presto-hive/src/test/java/com/facebook/presto/hive/parquet/AbstractTestParquetReader.java:1535]: duplicate line?

**Review by zhenxiao — CHANGES_REQUESTED** [review]
some testcase failures. could you please take a look?

**Comment by ClarenceThreepwood** [comment]
Thanks for the reviews @shixuan-fan and @zhenxiao. This should be all set to merge now

**Review by zhenxiao — APPROVED** [review]

**Pull request merged**